staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
 	})
 }
 
+func etagFor(data []byte) string {
+	return fmt.Sprintf("\"%X\"", sha512.Sum512(data))
+}
+
 // downloadOpenAPISpec downloads openAPI spec from /swagger.json endpoint of the given handler.
-func (s *openAPIAggregator) downloadOpenAPISpec(handler http.Handler) (*spec.Swagger, error) {
+// httpStatus is only valid if err == nil
+func (s *openAPIAggregator) downloadOpenAPISpec(handler http.Handler, etag string) (returnSpec *spec.Swagger, newEtag string, httpStatus int, err error) {
 	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
 	handler = request.WithRequestContext(handler, s.contextMapper)
 	handler = http.TimeoutHandler(handler, specDownloadTimeout, "request timed out")
 
 	req, err := http.NewRequest("GET", "/swagger.json", nil)
 	if err != nil {
-		return nil, err
+		return nil, "", 0, err
+	}
+	if len(etag) > 0 {
+		req.Header.Add("If-None-Match", etag)
 	}
 	writer := newInMemoryResponseWriter()
 	handler.ServeHTTP(writer, req)
 
 	switch writer.respCode {
+	case http.StatusNotModified:
+		if len(etag) == 0 {
+			return nil, etag, http.StatusNotModified, fmt.Errorf("http.StatusNotModified is not allowed in absense of etag")
```

Comment:
Does this mean a server with an unexpected `StatusNotModified` will kill our openapi aggregation?